code/game/objects/items/granters.dm

```diff
@@ -115,16 +115,50 @@
 ///SPELLS///
 
 /obj/item/book/granter/spell
-	var/spell
+	var/obj/effect/proc_holder/spell/spell
 	var/spellname = "conjure bugs"
+	var/spell_level = 0
+
+/obj/item/book/granter/spell/Initialize()
+	. = ..()
+	GLOB.spellbooks += src
+	update_name()
+
+/obj/item/book/granter/spell/Destroy()
+	. = ..()
+	GLOB.spellbooks -= src
+
+/obj/item/book/granter/spell/proc/update_name()
+	var/obj/effect/proc_holder/spell/S = new spell
+	S.spell_level = spell_level
+	S.nameSpell()
+	name = "spellbook of [S.name]"
+	qdel(S)
+
+/obj/item/book/granter/spell/proc/level_up_book(amount, message)
+	if(spell_level + amount > spell.level_max)
```

Comment:
this doesn't do anything. i don't think it can get level_max without creating a new spell to get it from